### PR TITLE
feat(upgrade-job): add support for loki-stack helm chart upgrade

### PR DESCRIPTION
**Pull request by niladrih**
Requires #386.

Overview:
- adds new yq commands to delete objects, append to objects, and append to arrays.
- adds special upgrade values.yaml update for loki-stack changes from #386. Introduces migrations and set_literal_value updates to alter the existing yaml till it resembles the new one.
- replaces all `&Path` is the `helm::yaml::yq` module with `AsRef<Path>`.

Changes:
- removes .loki-stack.loki.config.ingester.lifecycler.ring.kvstore as it
does not exist in loki-stack v2.9.11
- adds set value for the grafana/loki container image tag
- removes .loki-stack.promtail.config.snippets.extraClientConfigs as it
does not exist in loki-stack v2.9.11
- removes .loki-stack.promtail.initContainer as it does not exist in loki-stack
v2.9.11
- migrates .loki-stack.promtail.config.lokiAddress to .loki-stack.promtail.config.clients
- adds yq command_output helper function
- add set value for .loki-stack.promtail.readinessProbe.httpGet.path
- migrate .loki-stack.promtail.config.snippets.extraClientConfigs to
.loki-stack.promtail.config.clients

### CODE REVIEW

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by tiagolobocastro** [comment]
> Replaces the `loki` and `promtail` yaml objects entirely to new defaults. Doesn't honour user's configuration for the older loki/promtail. This should be okay, as this is not a critical storage component (@tiagolobocastro wdyt?).

I don't think this is a good idea, enabled and persistence are in loki, we should not modify these.

**Comment by niladrih** [comment]
> > Replaces the `loki` and `promtail` yaml objects entirely to new defaults. Doesn't honour user's configuration for the older loki/promtail. This should be okay, as this is not a critical storage component (@tiagolobocastro wdyt?).
> 
> I don't think this is a good idea, enabled and persistence are in loki, we should not modify these.
> 

The enabled is in `loki-stack`. This is preserved. There are 4 items under loki-stack: enabled, localpvScConfig, loki, promtail. The loki and promtail are overwritten. The other 2, are retained.

**Comment by niladrih** [comment]
I agree that persistence block items should be taken forward.

**Comment by niladrih** [comment]
This change now handles loki-stack helm chart update like it handles other changes to helm chart default values: it sets the new values to the final values manually.
This means all user configuration is carried forward.

File: k8s/upgrade/src/bin/upgrade-job/helm/values.rs [hunk 1]
```diff
@@ -203,3 +249,70 @@ where
...
+        promtail_config_client,
+        upgrade_values_filepath,
+    )?;
+    // Merge the extraClientConfigs from the promtail v3.11.0 chart to the v6.13.1 chart's
+    // config.clients block. Ref: https://github.com/grafana/helm-charts/issues/1214
+    // Ref: https://github.com/grafana/helm-charts/pull/1425
+    if !target_values.promtail_extra_client_configs().is_empty() {
+        // Converting the YAML to a JSON because the yq command goes like this...
+        // yq '.config.clients[0] += {"tenant_id": "1", "basic_auth": {"username": "loki",
+        // "password": "secret"}}' file
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/helm/values.rs:286]: Don't understand this comment..
Did you mean to say yq works with a json syntax? If so it'd be clearer to just say that..

File: k8s/upgrade/src/bin/upgrade-job/helm/yaml/yq.rs [hunk 2]
```diff
@@ -169,9 +251,10 @@ impl YqV4 {
     }
 
     /// This sets in-place yaml values in yaml files.
-    pub(crate) fn set_value<V>(&self, key: YamlKey, value: V, filepath: &Path) -> Result<()>
+    pub(crate) fn set_literal_value<V, P>(&self, key: YamlKey, value: V, filepath: P) -> Result<()>
     where
         V: Display + Sized,
+        P: AsRef<Path>,
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/helm/yaml/yq.rs:257]: why so many AsRef<Path> vs &Path?

**Review by tiagolobocastro — APPROVED** [review]

File: k8s/upgrade/src/bin/upgrade-job/helm/yaml/yq.rs (hunk 2 above)
Comment (niladrih) [reply k8s/upgrade/src/bin/upgrade-job/helm/yaml/yq.rs:257]: There were AsRef<Path>s and &Paths in this module. Made all of the args AsRef<Path>.

Comment (tiagolobocastro) [reply k8s/upgrade/src/bin/upgrade-job/helm/yaml/yq.rs:257]: kind of pointless here IMHO as paths returns &Path anyway, and need to add extra lines for P: and as_ref() throughout

**Review by dsharma-dc — APPROVED** [review]

File: k8s/upgrade/src/bin/upgrade-job/helm/yaml/yq.rs (hunk 2 above)
Comment (niladrih) [reply k8s/upgrade/src/bin/upgrade-job/helm/yaml/yq.rs:257]: Alright, yeah that makes sense. These aren't standalone libraries anyway. Yup.

**Comment by niladrih** [comment]
Requires #401.

**Comment by niladrih** [comment]
bors merge

**Pull request merged**